### DISCUSSION TITLE
Parameter / DelegateParameter value validation

### DISCUSSION THREAD

**Discussion by Junliang-Wang** [microsoft/Qcodes#4819] [Ideas]
Hi!

I love the safety limits for Parameters, but I have a tiny suggestion for the value validation.
The problem is that DelegateParameter.validate does not check its source parameter limits.
Let's see the following case:
```python
from qcodes import Parameter, DelegateParameter
from qcodes.validators import Numbers

param = Parameter('param', set_cmd=None, get_cmd=None, vals=Numbers(-5,5))
delegated = DelegateParameter('delegated', source=p1)

# delegated(-10) --> it raises an error as expected
# delegated.validate(-10) --> it does not raise an error, but I think it should!
```
One solution is to add in .validate also the validation for the source parameter.
For instance, I have a simple method to validate a list of values for any qcodes parameter.

```python
def get_param_sources(param):
    l = []
    while hasattr(param, 'source'):  # DelegateParameter can also be delegated
        param = param.source
        l.append(param)
    return l

def validate_param_values(param, values):
    sources = get_param_sources(param)
    all_params = [param] + sources
    [pi.validate(vi) for vi in values for pi in all_params]
``` 
Although this may seem to be an edge case, it is very useful when one wants to validate all the sweeping values before running the actual measurement.
The problem with validating at the moment of setting the value is that the measurement can stop in the middle of the sweep, unless you have an exception block to handle the error.
Therefore, I think it is better to add also the validation of all the source parameters.

Cheers,
Junliang

**Comment by jenshnielsen** [comment]
Hi Junliang, 

Thanks for your question. That sounds like a reasonable ask. 
It should be relatively straightforward to implement a validator in the DelegateParameter class that
calls validate on the source parameter too. Would you be willing to give it a try to implement such a thing? I would be happy to review that.

**Reply by Junliang-Wang** [reply]
Hi Jens,

thanks for the reply.
Sure, I am glad to give it a try.
